src/VisualStudio/Core/Def/ProjectSystem/MetadataReferences/VisualStudioMetadataReferenceManager.cs

```diff
@@ -149,89 +149,90 @@ internal Metadata GetMetadata(string fullPath, DateTime snapshotTimestamp)
             GC.KeepAlive(newMetadata);
 
             return metadata;
-        }
 
-        /// <exception cref="IOException"/>
-        /// <exception cref="BadImageFormatException" />
-        private AssemblyMetadata CreateAssemblyMetadataFromTemporaryStorage(
-            FileKey fileKey, List<TemporaryStorageService.TemporaryStreamStorage> storages)
-        {
-            var moduleMetadata = CreateModuleMetadataFromTemporaryStorage(fileKey, storages);
-            return CreateAssemblyMetadata(fileKey, moduleMetadata, storages, CreateModuleMetadataFromTemporaryStorage);
+            // <exception cref="IOException"/>
+            // <exception cref="BadImageFormatException" />
+            AssemblyMetadata CreateAssemblyMetadataFromTemporaryStorage()
+            {
+                var moduleMetadata = CreateModuleMetadata(key);
+                return CreateAssemblyMetadata(key, moduleMetadata, CreateModuleMetadata);
+
+                ModuleMetadata CreateModuleMetadata(FileKey moduleFileKey)
+                {
+                    GetMetadataFromTemporaryStorage(moduleFileKey, out var storage, out var metadata);
+                    storages.Add(storage);
+                    return metadata;
+                }
+            }
         }
 
-        private ModuleMetadata CreateModuleMetadataFromTemporaryStorage(
-            FileKey moduleFileKey, List<TemporaryStorageService.TemporaryStreamStorage>? storages)
+        private void GetMetadataFromTemporaryStorage(
+            FileKey moduleFileKey, out TemporaryStorageService.TemporaryStreamStorage storage, out ModuleMetadata metadata)
         {
-            GetStorageInfoFromTemporaryStorage(moduleFileKey, out var storage, out var stream);
+            GetStorageInfoFromTemporaryStorage(moduleFileKey, out storage, out var stream);
 
             unsafe
             {
                 // For an unmanaged memory stream, ModuleMetadata can take ownership directly.
-                var metadata = ModuleMetadata.CreateFromMetadata((IntPtr)stream.PositionPointer, (int)stream.Length, stream.Dispose);
-
-                // hold onto storage if requested
-                storages?.Add(storage);
```

Comment:
instead of this choice here, the caller decides what to do with the storage instance.